packages/docs/versioned_docs/version-2.0.0-alpha-3/protocol/01-users-and-attesters.md

```diff
@@ -2,18 +2,16 @@
 
 ## Actors in Unirep
 
-### User 👤
+### Users 👤
+**receive reputation** from attesters and are able to **prove received reputation**.
 
-There are users who **receive reputation** and **prove received reputation**.
+* Users sign up by calling `userSignUp` on the Unirep contract with a signup proof.
+* Users' `identityCommitment` is revealed at this time and recorded in the contract to prevent double signup.
```

Comment:
I am confused by this wording.. where is the identity commitment "revealed"? Does this mean that it is exposed? Does the user ever see it? How is the identity commitment stored by the user so they can be recognized by their attester application?
What are the values being passed in to generate the identity commitment (nullifier and trapdoor)? Where do they come from / how do they relate to the user that prevents double signup?

Reply:
In the [`SignupProof`](https://developer.unirep.io/docs/circuits-api/circuits#signup-proof) you can see the `identityCommitment` as a public output. This is given to the blockchain with the proof data. The `identityCommitment` can also be thought of as the semaphore identity public key. They user can calculate it by hashing their private key.

The nullifier and trapdoor are secure random values generated by the user. The same as generating an Ethereum private key, just random data that the user stores/keeps secret.

The `identityCommitment` is `H(H(identityNullifier, identityTrapdoor))`. So if the `identityCommitment` is different the private values must also be different. This exposes an interesting attack though. We typically use the `identityNullifier` as the private input. This exposes an attack where the user changes only the trapdoor and not the nullifier, which would allow state tree leaf collisions, epoch key collisions, etc. Documented in #273